### DISCUSSION TITLE
How the the average time is calculated in the service graph

### DISCUSSION THREAD

**Discussion by maheddin092** [grafana/tempo#4382] [Q&A]
Here’s a corrected and polished version of your statement:

I am implementing traces in my application using the OpenTelemetry C++ SDK. I am facing an issue with service graph creation.

Consider there are two threads: "Receive Thread" and "Worker Thread". I am creating spans for operations in both threads. The Receive Thread receives a message and forwards it to the Worker Thread. However, in the service graph, the average time calculation only considers the span from the Receive Thread. It does not include the span from the Worker Thread.

![service graph](https://github.com/user-attachments/assets/b15fb321-bdfb-439d-ba47-6ec030370100)

In the image above, the red box represents the span from the Receive Thread, and the green box represents the span from the Worker Thread. However, in the service graph, only the span from the Receive Thread is being included in the calculations. How can I combine these two spans so that both are considered in the calculations related to the service graph?

**Comment by joe-elliott** [comment]
This is showing an entry point into your system as recorded by `Recieve span` from the `DMADM` service. It has the correct value. Your other connections are not showing. Please consult the metrics generator troubleshooting guide:

https://grafana.com/docs/tempo/latest/troubleshooting/metrics-generator/

and confirm that your span kinds are correctly set to "Server" and "Client"

There can also be some shenanigans surrounding the way PromQL treats the first value of a rated metric. We have recently been patching up different metrics in the generator to work around Prometheus's behavior. Perhaps try pushing 5 - 10 traces.
